Mathlib/RingTheory/Unramified/Finite.lean

```diff
@@ -0,0 +1,277 @@
+  have ⟨f, hf⟩ : ∃ (a : I →₀ S), elem R S = a.sum (fun i x ↦ x ⊗ₜ b i) := by
+    let b' := ((Basis.singleton PUnit.{1} S).tensorProduct b).reindex (Equiv.punitProd I)
+    use b'.repr (elem R S)
+    conv_lhs => rw [← b'.total_repr (elem R S), Finsupp.total_apply]
+    congr! with _ i x
+    simp [b', Basis.tensorProduct, TensorProduct.smul_tmul']
+  constructor
+  -- I claim that `{ fᵢbⱼ | i, j ∈ s }` spans `S` over `R`.
+  use Finset.image₂ (fun i j ↦ f i * b j) f.support f.support
+  rw [← top_le_iff]
+  -- For all `x : S`, let `bᵢx = ∑ aᵢⱼbⱼ`.
+  rintro x -
+  let a : I → I →₀ R := fun i ↦ b.repr (b i * x)
+  -- Consider `F` such that `fⱼx = ∑ Fᵢⱼbⱼ`.
+  let F : I →₀ I →₀ R := Finsupp.onFinset f.support (fun j ↦ b.repr (x * f j))
+    (fun j ↦ not_imp_comm.mp fun hj ↦ by simp [Finsupp.not_mem_support_iff.mp hj])
+  have hG : ∀ j ∉ (Finset.biUnion f.support fun i ↦ (a i).support),
+      b.repr (f.sum (fun i y ↦ a i j • y)) = 0 := by
+    intros j hj
+    simp only [Finset.mem_biUnion, Finsupp.mem_support_iff, ne_eq, not_exists, not_and,
+      not_not] at hj
+    simp only [Finsupp.sum]
+    trans b.repr (f.support.sum (fun _ ↦ 0))
+    · refine congr_arg b.repr (Finset.sum_congr rfl ?_)
+      simp only [Finsupp.mem_support_iff]
+      intro i hi
+      rw [hj i hi, zero_smul]
+    · simp only [Finset.sum_const_zero, map_zero]
+  -- And `G` such that `∑ₛ aᵢⱼfᵢ = ∑ Gᵢⱼbⱼ`, where `aᵢⱼ` are the coefficients `bᵢx = ∑ aᵢⱼbⱼ`.
+  let G : I →₀ I →₀ R := Finsupp.onFinset (Finset.biUnion f.support (fun i ↦ (a i).support))
+    (fun j ↦ b.repr (f.sum (fun i y ↦ a i j • y)))
+    (fun j ↦ not_imp_comm.mp (hG j))
+  -- Then `∑ Fᵢⱼ(bⱼ ⊗ bᵢ) = ∑ fⱼx ⊗ bᵢ = ∑ fⱼ ⊗ xbᵢ = ∑ aᵢⱼ(fⱼ ⊗ bᵢ) = ∑ Gᵢⱼ(bⱼ ⊗ bᵢ)`.
+  -- Since `bⱼ ⊗ bᵢ` forms an `R`-basis of `S ⊗ S`, we conclude that `F = G`.
+  have : F = G := by
+    apply Finsupp.finsuppProdEquiv.symm.injective
+    apply (Finsupp.equivCongrLeft (Equiv.prodComm I I)).injective
+    apply (b.tensorProduct b).repr.symm.injective
+    simp only [Basis.repr_symm_apply, Finsupp.coe_lsum, LinearMap.coe_smulRight,
+      LinearMap.id_coe, id_eq, Basis.tensorProduct_apply, Finsupp.finsuppProdEquiv,
+      Equiv.coe_fn_symm_mk, Finsupp.uncurry, map_finsupp_sum,
+      Finsupp.total_single, Basis.tensorProduct_apply, Finsupp.equivCongrLeft_apply,
+      Finsupp.total_equivMapDomain, Equiv.coe_prodComm]
+    rw [Finsupp.onFinset_sum, Finsupp.onFinset_sum]
+    simp only [Function.comp_apply, Prod.swap_prod_mk, Basis.tensorProduct_apply]
+    have : ∀ i, ((b.repr (x * f i)).sum fun j k ↦ k • b j ⊗ₜ[R] b i) = (x * f i) ⊗ₜ[R] b i := by
+      intro i
+      simp_rw [Finsupp.sum, TensorProduct.smul_tmul', ← TensorProduct.sum_tmul]
+      congr 1
+      exact b.total_repr _
+    trans (x ⊗ₜ 1) * elem R S
+    · simp_rw [this, hf, Finsupp.sum, Finset.mul_sum, TensorProduct.tmul_mul_tmul, one_mul]
+    · rw [← one_tmul_mul_elem, hf, finite_of_free_aux]
+      rfl
+    · intro _; simp
+    · intro _; simp
```

Comment:
```suggestion
    · intro; simp
    · intro; simp
```
You don't need them so I'm not sure there's any point in putting them, like you wouldn't put `apply foo _`.